_posts/2020-02-17-constructor-injection.md

```diff
@@ -0,0 +1,189 @@
+
+Constructor injection is extremely useful since we do not have to write separate business logic everywhere to check if the all the required dependencies are loaded thus simplifying code complexity.
+
+## Identifying Code smells
+
+Constructor injection can help us identify if our bean is dependent many other classes. If our constructor has large number of arguments, then we can be sure that our class has too many responsibilities. It is not a sign of quality code implying it should be refactored to better address proper separation of concerns.
+
+Constructor injection can help us find out circular dependencies in the code. If it exists, Spring would throw `BeanCurrentlyInCreationException` . This can help the developer identify classes that have circular dependency.
+
+## Preventing NullPointerException
+
+Constructor injection simplifies writing unit test cases for any spring application. Using `Mockito`, we can create complete mock objects with required dependencies.
+
+You may ask why can't I mock objects created with setter injection. **Mockito unlike Spring fails with a null pointer exception if it cannot inject any of the dependencies**.
+
+Let us consider an example test case for Cake class that uses setter injection. Here the test works fine. 
+```java
+@RunWith(SpringJUnit4ClassRunner.class)
+@SpringBootTest(classes = ExampleApplicationSI.class)
+public class TestCakeClassSetterInjection {
+
+	@Autowired
+	Cake cake;
+
+	@Test
+	public void testSetterInjection() {
+		String testColor = cake.getFlavor().getColor();
+		Assert.assertEquals(testColor, " White ");
+		String toppingsName=cake.getToppings().getToppingName();
+		
+		//check if the dependency is not null
+		if(toppingsName!=null) {
+			Assert.assertEquals(toppingsName, "gems");
+		}
+		
+		...
+}
+``` 
+In future if someone adds additional dependencies to this class, the unit test silently fails with a null pointer exception. It is very difficult to find the reason because the new dependencies is not within the visibility for Cake’s client classes.
+
+Let us consider the Cake class example created using constructor injection.
+```java
+@RunWith(SpringJUnit4ClassRunner.class)
+@SpringBootTest(classes = ExampleApplicationCI.class)
+public class TestCakeClassConstructorInjection {
+
+	@Autowired
+	Cake cake;
+	
+	@Test
+	public void testConstructorInjection() {	
+		String testColor=cake.getFlavor().getColor();
+		Assert.assertEquals(testColor, " White ");
+	}
+	
+	...
+}
+```
+
+Now when we need a new dependency, we add it to the constructor. Thus, whenever we create an object of this class, we make sure that all its dependencies are present. This is a big gain since we do not have test with half created objects.
```

Comment:
Similar to above, I don't really understand what you're saying here. Please try to explain in very simple code examples and some text below them.